### PR TITLE
Add plugin type

### PR DESCRIPTION
**Pull request by ndmitchell**

### CODE REVIEW

File: src/Development/IDE/Plugin.hs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import qualified Language.Haskell.LSP.Core as LSP
+import Language.Haskell.LSP.Messages
+
+
+data Plugin = Plugin
+    {pluginRules :: Rules ()
+    ,pluginHandler :: PartialHandlers
+    }
+
+instance Default Plugin where
```
Comment (cocreature) [inline src/Development/IDE/Plugin.hs]: Not a big fan of `Data.Default`. Can we just export `defPlugin` instead?

File: src/Development/IDE/Plugin.hs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    def = Plugin mempty def
+
+instance Semigroup Plugin where
+    Plugin x1 y1 <> Plugin x2 y2 = Plugin (x1<>x2) (y1<>y2)
+
+instance Monoid Plugin where
+    mempty = def
+
+
+codeActionPlugin :: (LSP.LspFuncs () -> IdeState -> TextDocumentIdentifier -> Range -> CodeActionContext -> IO [CAResult]) -> Plugin
```
Comment (cocreature) [inline src/Development/IDE/Plugin.hs]: Why is `codeActionPlugin` in here instead of `Development.IDE.Plugin.CodeAction`?

Comment (ndmitchell) [reply src/Development/IDE/Plugin.hs]: The idea is that `codeActionPlugin` is the general helper for writing code actions, so when there are two code action plugins, they both use this helper, but have different internals. It's closer to the model HIE follows. The aim would eventually for there to be 5 or 6 such helpers in this module.

File: src/Development/IDE/Plugin.hs (hunk 1 above)
Comment (ndmitchell) [reply src/Development/IDE/Plugin.hs]: We're already using Data.Default for the rest of this API. I wanted to stay consistent.

Comment (cocreature) [reply src/Development/IDE/Plugin.hs]: Fair enough

File: src/Development/IDE/Plugin.hs (hunk 2 above)
Comment (cocreature) [reply src/Development/IDE/Plugin.hs]: Makes nense.

**Pull request merged**